Implement Glimmer Engine

wycats opened the pull request:
Glimmer is the next-generation rendering engine for Ember.js. Built on top of HTMLBars, streams, and an entirely rewritten view layer, Glimmer offers the best performance of any of the major JavaScript frameworks while remaining backwards compatible with Ember 1.x apps.

---

The Virtual DOM showed the world that render-time diffing delivers awesome performance by minimizing DOM updates and keeping unchanged elements stable. But like any abstraction, it comes with a cost. In this case, constructing the virtual DOM to compare requires at least some subset of components do a full re-render, and building virtual DOM nodes for static areas of your markup that will never change for every change requires many more allocations and increases GC pressure.

Glimmer analyzes templates at compile time, relying on the fact that Handlebars' declarative syntax clearly differentiates between static areas, which make up the majority of a template, and the dynamic areas that can change. Instead of invoking a `render()` method and diffing the result, Glimmer only walks the (much smaller) existing dynamic tree.

---

Handlebars' declarative syntax means that from the perspective of an app, the template is being "re-rendered every time", but we can take advantage of static knowledge to reduce the work that we need to do. User code, like helpers and computed properties, are executed during the walk, but only to get (primitive) **values** which can be `===` compared, not to build up a new tree. **In other words, the programming model is equivalent to "render every time", but we take advantage of the declarative nature of Ember's APIs to reduce work**.

Internally, instead of creating a virtual DOM, Glimmer builds a tree of streams, each stream pointing to a node in the DOM. On initial render, we track the last value (always a primitive) that we inserted into the DOM. The streams are not exposed to application code; Glimmer executes user code when necessary and pushes the new value into the stream.

When re-rendering, we walk the tree and flush the streams. If the primitive value produced by the stream has not changed, we do nothing. If it has changed, we write the change to the DOM. **Like Virtual DOM approaches, this is a "write-only" algorithm**.

One of the benefits of this approach is that **we can naturally fuse the Ember/Polymer model of observing individual properties with the React model of explicitly re-rendering areas of the template**. No matter what happens, the process of revalidation walks the tree, looking for dirty nodes, and revalidates any dirty nodes it finds.

The only difference is how much of dynamic tree is marked dirty before revalidation begins. Also, because observation can only dirty dynamic nodes (and schedule a top-down revalidation), multiple observers and re-renders that occur during a single run-loop still only produce **a single batch of DOM updates**.

In essence, the biggest difference between Glimmer and traditional virtual DOM approaches is that we diff values, not DOM. This approach not only lets us avoid the algorithmic complexity of tree diffing (instead, we just `===` check two values), it also lets us avoid doing many render-time allocations.

[Slides from the talk at EmberConf](http://f.cl.ly/items/0t031v2Z3y001V1N0F3N/Virtual%20DOM.pdf)
### Work List

(this looks like more things than it actually is)
- [x] merge subscription change into our branch
- [x] linkRenderNode can mark stability
- [x] Ember should return true from `linkRenderNode`
- [x] make sure all child streams correctly disconnect from parents
- [x] add stream pruning logic that correctly prunes ancestors
- [x] add scope teardown hook to HTMLBars (recursive upon `clear` or `remove`)
  - [ ] this is also responsible for view layer teardown
- [x] Propagate correct visitor through descendent hooks (so dirty checks can be avoided for a subtree)
- [x] Do not dirty check render nodes that were just created (fixes regression)
- [x] Add "classify" hook to decide which hook `content` et al should delegate to
- [x] Make sure that #view is idempotent
- [x] Components and views should create a new scope frame
  - [x] Implement custom scope in Ember (so that ambient `view`, `controller`, etc. can be stored)
  - [x] Eliminate `*component*` hacks
- [x] Lifecycle hooks on the renderer
- [x] Move code duplicated between `#view`, `#component` and `Ember.View` into the Renderer
- [ ] Support reflecting attributes onto angle-bracket components
- [ ] Pass mutable bindings as "mutator" objects
- [ ] Support `foo={{action "bar"}}`
- [ ] Legacy support: Implement unknownProperty/setUnknownProperty to work with mutator objects
  - [ ] Deprecate setting non-`mut` attributes, and support `(mut foo)` sexprs in legacy Handlebars syntax
- [ ] Investigate executing teardown callbacks at a "quiescent" state ("vent plasma")
  - [ ] Make sure to consider the rpflo stress test situation
- [x] Implement a `zip` and `map` abstraction for streams to clean up ad hoc stream creation
- [ ] Add tests for memory leaks
- [ ] Make sure that custom render functions that mutate the buffer continue to work
- [ ] `options.template` should not exist if there is no template
- [ ] feature flag all the things
- [ ] Handlebars helper compat
- [ ] Template function that returns a string (legacy)
- [X] Context shifting `#each`
- [ ] `collection` helper
- [ ] `{{input}}`
  - [X] `TextField`
  - [X] `Checkbox`
  - [ ] `TextArea`
- [ ] `Ember.Select`
- [ ] `unbound`
  - [X] non-block form
  - [ ] block form (partially done, but not completely)
- [X] `#view` and `view` helper
- [ ] `#view` and `view` helper with back-compat edge cases
- [X] `#with` helper
- [ ] `with` helper with back-compat edge-cases
- [ ] `yield` without a block
- [ ] `{{render}}`
- [ ] `ContainerView`
- [X] Top-level view APIs
  - [X] `createElement`
  - [X] `appendTo`
  - [X] `replaceIn`
  - [ ] `destroyElement`
- [ ] `isVisible`

wycats commented:
React Hook | Ember Hook | Server? | Initial Render? | Rerender? | Purpose
--- | --- | --- | --- | --- | ---
`componentWillMount` | `init` | Yes | Yes | No | Set initial component state without triggering re-render
`componentDidMount` | `didInsertElement` | No | Yes | No | Provides opportunity for manual DOM manipulation
`componentWillReceiveProps` | `willReceiveAttrs` | No | No | Yes | React to changes in component attributes, so that `setState` can be invoked before render
`shouldComponentUpdate` | Maybe N/A | No | No | Yes | Gives a component an opportunity to reject downstream revalidation
`componentWillUpdate` | `willUpdate` | No | No | Yes | Invoked before a template is re-rendered to give the component an opportunity to inspect the DOM before updates have been applied ([example](http://blog.vjeux.com/2013/javascript/scroll-position-with-react.html))
`componentDidUpdate` | `didUpdate` | No | No | Yes | Invoked after a template is re-rendered to give the component an opportunity to update the DOM ([example](http://blog.vjeux.com/2013/javascript/scroll-position-with-react.html))
`componentWillUnmount` | `willDestroyElement` | No | No | Yes | The inverse of `componentDidMount`/`didInsertElement`; clean up anything set up in that hook
[N/A](https://github.com/facebook/react/issues/2074) | `willRender` | No | Yes | Yes | In Ember, executed *both* after `init` and after `willUpdate`*
[N/A](https://github.com/facebook/react/issues/2074) | `didRender` | No | Yes | Yes | In Ember, executed *both* after `didInsertElement` and `didUpdate`*

* These hooks can be used in cases where the setup for initial render and subsequent re-renders is idempotent (e.g. `$().addClass`) instead of duplicating the logic in both places. In most cases, it is better to try to make these hooks idempotent, in keeping with the spirit of "re-render from scratch every time".

wycats commented:
The above comment is an attempt to classify the hooks in React and Ember so that we can make sure that the new Ember components implement *at least* the hooks that React folks have identified as being important in the programming model.

rwjblue commented:
All new API's will also need to be feature flagged.

wycats commented:
@rwjblue yessir

Linicks commented:
The init hook name is a little unclear compared to componentWillMount.  Could it be named initElelement, or something similar ?

stefanpenner commented:
> The init hook name is a little unclear compared to componentWillMount. Could it be named initElelement, or something similar ?

I would prefer leaving it as init, as it clearly describes what it handles, initialization of the object. Which is the same between all ember objects, introducing a deviation here seems strange.

stefanpenner commented:
@wycats @tomdale good job guys :)

alexspeller commented:
:heart_eyes:

alexdiliberto commented:
This is amazing :+1: :+1: :+1:

Linicks commented:
The problem with just using init is that init is used in so many places in the tech world.  In an Ember only world it would be clear and concise, but in an ever increasing complex tech stack it's nice to have that extra bit of clarity.  I think that's why FB chose an even more verbose naming scheme than my proposal.  I have had the chance/misfortune of maintaining allot of legacy software, and appreciate when the code is self describeing as practical.  This is really a preference in either direction, but after being in this game for so many years I'm getting a little opionated :)  Either way this is exciting stuff, and look forward to seeing this come together in Ember.

ebryn commented:
@Linicks if you look at their ES6 syntax, you'll see they're utilizing the constructor to initialize `this.state`

fivetanley commented:
What is a "ShadowRoot"?

gonvaled commented:
This being a "DOM write-only" implementation, how does ember deal with changes to the DOM done by third-party components, like jQuery based changes, which are not controlled by Ember?

davidlormor commented:
"the Dynamic Tree only includes nodes that may have changed (...) and never needs to compare static content that will never change"...sounds fast! :rocket: :+1:

stefanpenner commented:
> This being a "DOM write-only" implementation, how does Ember deal with changes to the DOM done by third-party components, like jQuery based changes, which are not controlled by Ember?

It doesn't and likely cannot support this.

mgenev commented:
@stefanpenner that means we can no longer use bootstrap in an ember app unless a bootstrap-for-ember version is written, right? Same goes for foundation etc...

stefanpenner commented:
> @stefanpenner that means we can no longer use bootstrap (jQiery dependent) in an ember app unless a bootstrap-for-ember version is written, right? Same goes for foundation etc...

No, this is no different then today. If you component is re-rendered your bootstrap and jquery code needs to re-run.

wycats commented:
@stefanpenner @mgenev I think there's a little bit of confusion here.

From new-world Ember's perspective, as long as the element for a component still exists in the DOM, it doesn't matter where it has been moved, whether it has been decorated, etc.

There are a few restrictions on moving elements around, but they're very minor and would be easy to abstract. Stay tuned as we continue to make more progress.

ef4 commented on packages/ember-htmlbars/tests/helpers/new_each_helper_test.js:
```diff
@@ -0,0 +1,334 @@
+// and generate:
+//
+// - {{#each p in people}} (legacy)
+// - {{#each people as |p|}} (legacy)
```
why is `{{#each people as |p|}}` marked as legacy? Are we not keeping that syntax?

ef4 commented on packages/ember-metal/lib/utils.js:
```diff
@@ -105,6 +105,14 @@ function intern(str) {
+  // want to require non-enumerability for this API, which
+  // would introduce a large cost.
+
+  return intern(debugName + ' [id=' + GUID_KEY + Math.floor(Math.random() * new Date()) + ']');
```
Are there really platforms where Math.random() is so broken that we need to multiply by `new Date`?

tomdale replied:
Dunno, let's have our IE friends run http://jsbin.com/lexonidara/2/edit?html,js,output overnight.

tomdale replied:
Our rationale here was that `Math.random()` does not have cryptographically-guaranteed entropy, and the result of a collision would be disastrous and also impossible to debug.

Our goal was to minimize the odds of collision by introducing a constantly increasing factor (i.e., time), such that the odds plummet if the symbol generator is not called in the same millisecond. On reflection, though, perhaps we meant to sum the two—multiplying current millis by the result of `Math.random()` reduces the odds of collision but it doesn't do much to provide stronger guarantees around avoiding collisions if the result of `Date.now()` changes.

ef4 replied:
I think you're actually increasing the odds of collision significantly, not reducing them.

"Not cryptographically strong" just means that it's possible to make better-than-expected guesses about future values based on past values. It doesn't mean the distribution is skewed. `Math.random()` is supposed to give a reasonably uniform distribution over 64 bits.

When you multiply by date and cast down to integer, you throw away roughly 24 bits worth of entropy, because the date is only going to be roughly `2^40`. So you end up with only the integers integers from 0 to 2^40.

If instead you did roughly:

    var MAX_DOUBLE = Math.pow(2, 64);
    return Math.floor(Math.random() * MAX_DOUBLE);

you will get integers from 0 to 2^64.

jdalton replied:
Ran this in IE11 overnight and no collisions.

loopmode replied:
I came from using millis to multiplying by random in that particular order, so the idea was to allow for more possible values during a single CPU cycle. Maybe the author had a similar purpose in mind.

chadhietala commented on packages/ember-htmlbars/lib/env.js:
```diff
@@ -2,35 +2,70 @@ import environment from "ember-metal/environment";
+import componentKeyword from "ember-htmlbars/keywords/component";
+import partial from "ember-htmlbars/keywords/partial";
+
+merge(emberHooks.keywords, {
```
Currently there are no upstream keywords in HTMLBars, so I believe this needs to be:
```
emberHooks.keywords = {};
merge(emberHooks.keywords, {...});
````

rwjblue commented:
Updated title + description based on EmberConf presentation.

toblender commented:
Minor correction: teack === track?

morenoh149 commented:
@toblender yeah. I thought he meant `teach`

darkbaby123 commented:
@wycats You mentioned a lot about diffing values to check if they're dirty. I'm curious what's the difference between Glimmer's diffing logic and Angular's dirty-checking.

As I know Angular dirty-checking also checks values (expressions) one by one and execute corresponding callbacks (mostly change doms). But:
1. Angular doesn't have run loop so it's can not batch update doms.
2. watchers need to be triggered by `$apply()` manually sometimes.

drogus commented:
> I'm curious what's the difference between Glimmer's diffing logic and Angular's dirty-checking.

These two diffing situations are not happening at the same level. Angular's diffing happens outside of a view layer in order to check if a template change is needed. So let's say that you have a `user = { login: 'drogus' }` object. And you use it in a template as `<span class="login">{{user.login}}</span>`. Now, in order to update the template automatically you need to know if a value of the `user` object changed. In Angular it's done by dirty checking all of the properties used in templates. In Ember properties are observed, so it knows when a property is changed.

This PR and diffing described here happen in a view layer when a framework already knows that it needs to rerender something, but it needs to figure out what to rerender exactly. DOM is slow, so regenerating the entire template may be costly and here comes the diffing algorithm that React started with.

bitinn commented:
One interesting side-effects of React (or other similar dom-diffing) approach is you can worry less about what's being updated in DOM (as oppose to update them manually on value change).

What sort of limit will value-diffing approach impose on DOM manipulation? Would you say this approach is closer to something like [Riot](https://muut.com/riotjs/compare.html#string-based-vs-dom-based) than React?

mixonic commented:
@bitinn your understanding is pretty good. I think the comparison to Riot is a good one.

Your points one and two are correct IMO. Point three is a bit off, but I'm not sure what you are thinking when you say "traditional template renderer". A full rerender in glimmer will re-use the static DOM anywhere the template being used has not changed, and will only update dynamic content if it has changed. Where the template has changed, Glimmer will `cloneNode` a cached fragment of pre-hydrated DOM.

I'm sure the architecture will be discussed in more details as this gets closer to landing.

jonathanKingston commented:
Is the intention to swap out `ShadowRoot` with `document.ShadowRoot` eventually?

mmun committed:
Split out Dependency and Subscriber classes

mmun committed:
Refactor stream dependencies

Stream dependencies now mock the Stream interface which accomodates some
leaner code.

Stream now has a revalidate hook which is used for manual dependency
management. This hook is invoked in two cases:

  1. The stream just activated.
  2. The stream is already activate and is about to re-compute.

Commit:
Made more {{#view}} tests pass

Commit:
Attrs, scope and view destruction cleanup

This commit fixes several edge cases in how scope was managed in
templates, how attrs are looked up, and how views are cleaned up,
particularly the willDestroyElement user hook.

Commit:
Take machete to inBuffer state

Commit:
Implement {{collection}} helper

Commit:
Unpend passing collection tests

Commit:
Helper for finding pending tests that now pass

wycats committed:
Got 75 more tests passing in ember-htmlbars

wycats committed:
Left off a bunch of new files

wycats committed:
Get more tests passing

wycats committed:
Get more tests passing

mmun committed:
Fix style errors

mmun committed:
Mark 3 regressions as skipped

mmun committed:
Get action keyword mostly working

mmun committed:
Bump htmlbars version

rwjblue committed:
Update HTMLBars to v0.12.0.

No longer requires `npm link` for HTMLBars.

Commit:
Fix rebase conflicts and get tests passing

mixonic committed:
Port getViewClientRects, getViewboundingClientRect to Glimmer

mixonic committed:
Drop tests for removed isVirtual behaviors

mixonic committed:
Reenable tests related to undefined/null attr values

Commit:
Rerender {{with}} when param changes

Commit:
Always set shadowScope’s `view` property

Commit:
Fix yield in nested views

rwjblue committed:
Use `isStream` helper to check for streamness.

rwjblue committed:
Prevent errors in updateScope hook when view is null.

rwjblue committed:
Ensure Ember.Handlebars.makeBoundHelper helpers get `options.data.properties`.

mmun committed:
Get render keyword mostly working

rwjblue committed:
Throw an assertion if `makeBoundHelper` is provided a block.

rwjblue committed:
Allow Handlebars compat helpers to specify additional dep keys.

In Ember.Handlebars helpers it is possible to add additional dependent
keys:

```javascript
Ember.Handlebars.helper('capitalize-name', function(value) {
  return get(value, 'name').toUpperCase();
}, 'name');
```

---

Unfortunately, the area that we know about the render node and can add
more dependent keys, had no knowledge of the individual helper being rendered.

Thsi commit adds a `findHelper` call to `linkRenderNode` so that we can
determine if additional dependent keys are needed. This is sub-optimal
because the additional `findHelper` may have a negative impact on render
speed, but a better solution is not currently possible.

rwjblue committed:
Process `fooBinding` in general.

Previously, the `TransformOldBindingSyntax` AST walker was only
processing `{{view}}` invocations. This change makes it process all
helper invocation, and updates a number of related tests.

In order to do that, we had to change the asserts and deprecations in
the transform so that when the `loc` information is not present no
errors are thrown.

mmun committed:
Fix debugger keyword

rwjblue committed:
Fix expected deprecations for recently fixed tests.

rwjblue committed:
Update to use new _itemViewTemplate for CollectionView test.

The `{{collection}}` helper sets this automatically, so we need to
ensure the tests follow suit.

rwjblue committed:
Fix Ember.LinkView export test.

rwjblue committed:
Add `QUnit.skip` to feature flagged tests.

rwjblue committed:
Use `expectAssertion` for testing Ember.assert.

rwjblue committed:
Get node tests running.

Some are still skipped, but will require further thought/details on
passing the renderer through to all children in
`ComponentNode.prototype.createOrUpdateComponent`.

rwjblue committed:
Remove unused import.

Fixes JSHint error.

rwjblue committed:
Enable LOG_VIEW_LOOKUPS for {{outlet}} keyword.

Note: that {{outlet}} templates without a view instance do not get a
default view any longer (the template is just rendered).

rwjblue committed:
Honor `viewName` property with `{{view}}` helper.

rwjblue committed:
Restore Component.prototype.targetObject.

This is needed to target the parentView.controller with the action (when
it is provided in the template).

It is possible that in 2.x this conceptually needs to change from
`parentView.controller` to something else, but the general idea will
still remain.

mmun committed:
Add deactivate hook in Stream, fix KeyStream observers lifecycle

The commit ensures observers for a KeyStream are removed correctly.
A new revalidate hook was introduced as a place to clean up any manual
dependencies. This hook is invoked when the last subscriber is removed
(including when the stream is being destroyed).

mmun committed:
Bump htmlbars to 0.12.1

rwjblue committed:
Make Handlebars compat helpers properly stamp themselves as HTMLBars.

rwjblue committed:
Fix expected deprecation for `attributeBindings`.

ef4 committed:
Fixing outlets inside {{render}} helper

Four new passing tests.

mmun committed:
Bump htmlbars to 0.13.0

This commit modifies the AST transforms to match the Handlebars 3.0 AST.

ef4 committed:
Fix {{render}}'s ownerView

And remove the workarounds from it being wrong.

rwjblue committed:
Add view._defaultTagName for overriding CollectionView tagName.

wycats committed:
Handle legacy View scope better

ef4 committed:
implement textarea keyword

Three more passing tests

ef4 committed:
drop unused argument

ef4 committed:
fix the {{outlet view="foo"}} form

ef4 committed:
Updating assertion language

I don't see a reason to bend over backwards to support the exact same
assertion language for this (deprecated) feature. This new message is
sufficiently clear.

ef4 committed:
drop unused code

ef4 committed:
support {{outlet viewClass=...}}

ef4 committed:
fix another outlet view test

ef4 committed:
fix another outlet assertion

ef4 committed:
deprecate {{outlet}} view and viewClass params

mitchlloyd committed:
Get further on classNameBindings

ef4 committed:
fix template context when using a 1.0-style template and controller

42 tests green

ef4 committed:
documenting non-obvious code

ef4 committed:
Dropping dead code

As pointed out by @mitchlloyd in #10871.

ef4 committed:
Merge PR #10871 from mitchlloyd/idempotent-rerender

mmun committed:
Fix unbound keyword

This commit also adds several tests for the path and helper usages of
the subexpression form, e.g.

```hbs
{{capitalize (unbound  foo)}}
```

and

```hbs
{{capitalize (unbound if foo bar baz)}}
```

rwjblue committed:
Add skipped test for compat mode attrs-proxy.

rwjblue committed:
Fix typo in compat mode attrs test.

tilde-engineering committed:
Update Ember.Select to use HTMLbars templates

tomdale committed:
Get Ember.Select tests passing

tomdale committed:
Set component props to attrs if defined

mmun committed:
Fix last textarea test

ef4 committed:
Merge branch 'master-before-tomhuda' into ir-before-tomhuda

Conflicts:
	packages/ember-routing-htmlbars/lib/helpers/render.js
	packages/ember-template-compiler/lib/plugins/transform-each-in-to-hash.js
	packages/ember-views/lib/attr_nodes/attr_node.js

ef4 committed:
prevent test suite explosion due to new AST transform options

ef4 committed:
drop incompatible options.inverse bugfix and leave skipped test

ef4 committed:
porting non-block-param {{#with}} deprecation to idempotent-rerender

ef4 committed:
post-merge cleanup

ef4 committed:
skipping tests that don't yet pass on glimmer

rwjblue committed:
Fix current node tests.

tilde-engineering committed:
Add note on Checkbox inheriting from Component

tilde-engineering committed:
Unify legacy class binding into AST preprocessor

There are several paths and microsyntaxes for creating class name
bindings for a component’s element via Handlebars. For example:

`{{view ‘my-view’ class=“hello” classBindings=“:foo isApp:is-an-app”}}`

With HTMLbars, these have been deprecated in favor of a single API:

`{{view class=“hello foo (if isApp ‘is-an-app’)”}}`

This commit rewrites the AST of templates that contain the old syntax
into the new syntax at template compile time.

tilde-engineering committed:
ContainerView should inherit from View

While it was a noble effort, the attempt to have ContainerView subclass
Component did not work because it broke propagation of the `controller`
template keyword.

tilde-engineering committed:
Implement collection’s itemClass

mitchlloyd committed:
Fix spacing between omitted class names

mitchlloyd committed:
Wrap `set` in run loop

rwjblue committed:
Update HTMLBars to 0.13.1.

Updates HTMLBars to ensure that initially undefined/null
`attributeBindings` are not set on the views element.

rwjblue committed:
Merge pull request #10886 from mitchlloyd/html-class-spacing

Fix class spacing for classNameBindings

rwjblue committed:
Enable tests that are now passing.

rwjblue committed:
Update link-render-node to handle missing params.

The `{{component}}` keyword specifies `null` as `params` (since a
component cannot receive ordered arguments).  This commit simply adds a
guard to ensure we do not attempt to access `params.length` when
`params` is `null`.

rwjblue committed:
Fix a couple Ember.Select tests.

mitchlloyd committed:
Fix isVisible on first render

ef4 committed:
dropping dead code: outlets don't use the rerender hook

mitchlloyd committed:
Get isVisible with container tests passing

mitchlloyd committed:
Add fix for overwriting style attribute bindings

mmun committed:
Fix inverse block in compat helpers

ef4 committed:
dropping dead code

this is all old implementation of pre-glimmer outlet behavior

ef4 committed:
rely on stable childEnvs to for outletState

ef4 committed:
fix use of template & layout together

ef4 committed:
provide deprecation warning if we fall back to template

mmun committed:
Remove invalid test

rwjblue committed:
Merge pull request #10888 from mitchlloyd/is-visible-fixes

isVisible fixes

rwjblue committed:
Add Ember specific AttrMorphClass to provide style warning.

mitchlloyd committed:
Use templates to test View.nearestTypeOf.

tilde-engineering committed:
Munge attrs before creating ComponentNode

tilde-engineering committed:
For backwards-compat, make attrs available at init

rwjblue committed:
Implement custom element hook to allow deprecating legacy unbound helpers.

rwjblue committed:
Correct assertion messages in tests.

mmun committed:
Fix a test using an undefined context

Context-changing helpers no longer allow an undefined context. Instead,
null is substituted for the context.

mmun committed:
Remove ember-metal-views tests

These tests were removed because they were only testing a stub renderer
defined in the tests.

rwjblue committed:
Remove unused AttrNode.

tilde-engineering committed:
Implement legacy {{each}} features

tilde-engineering committed:
WIP on {{each itemController}}

mixonic committed:
Completely remove the concept of isVirtual

MetamorphView (and the Metamorph mixin) are the final bastion of
virtual views. This commit removes much of their use.

MetamorphView had two behaviors: First, it was tagless. Second, it was
not included in the view hierarchy. The first concept (taglessness) is
still valid in Ember post-Glimmer. For this, we use `tagName: ''`.

The second concept (virtual views) is not longer present. So views which
previously used MetamorphView will not longer be able to escape the view
hierarchy. In most cases, the intent was likely to be tagless and being
virtual was simply a side effect.

OutletView and EachView will now be present in childViews and will be
parentViews.

`{{render` used to always create a view. Now, it will only create a view
if one is set by the user. By default it is only a componentNode and has
no view at all.

The final spot MetamorphViews are used it for the default itemView and
emptyView on EachView. This should be easy to address in a manner
similar to {{render: If the user does not specify a view, none should be
required.

After this, MetamorphView will no longer be used by internals.

rwjblue committed:
Merge pull request #10903 from mixonic/idempotent-rerender-no-virtual

Completely remove the concept of isVirtual

rwjblue committed:
Add missing -legacy-each registration for deprecated tests.

mmun committed:
Rewrite some tests to use templates instead of a buffer

mmun committed:
Reimplement hasBlock and hasBlockParams

mixonic committed:
Make createElement work without view entering inDOM state

createElement requires that a view be rendered without the lifecycle
hooks being rendered. This leaves it in the hasElement state instead of
moving to inDOM as a normal render would.

mixonic committed:
Clean up some lifecycle tests

mixonic committed:
Throw when appending and already appended view

mixonic committed:
Can presume the view.parentView is not virtual

ef4 committed:
fix tests that don't cleanup Application instances

ef4 committed:
drop dead code

this argument was used by the pre-glimmer OutletView only

ef4 committed:
topmost outlet must ensure its child has a DOM element

This keeps us consistent with the DOM structure we produced
pre-glimmer. The topmost out ensures that it's child (not itself) always
has a real View with an element.

ef4 committed:
fix action targeting when there is no view

ef4 committed:
this was already passing, not sure when forceskipped didn't find it sooner

ef4 committed:
refactor test to avoid buffer.push & appendChild

mmun committed:
Modernize child_views_test

Use templates instead of render method

mmun committed:
Use template instead of render buffer

mmun committed:
Reimplement willInsertElement

mmun committed:
Remove MetamorphView, salvage useful metamorph view tests

mmun committed:
Split up render_test.js

- Rename render_test.js to view_test.js.
- Move conatiner view related tests to container_view_test.js.
- Remove invalid tests that are overly specific about rerender
  behaviour. There are much more comprehensive tests in view_lifecycle_test.js.

mitchlloyd committed:
Get more tests with view#remove working

rwjblue committed:
Update HTMLBars.

rwjblue committed:
Update emberjs-build to use esperanto for HTMLBars packages.

rwjblue committed:
Update assertions to show NullLiteral type works properly.

mitchlloyd committed:
Cancel scheduled view#append on view#remove

If a view is in a ‘preRender’ state (or any state for that matter)
attempting to destroy the element will effectively make any pending
view#append operations a noop.

mitchlloyd committed:
Destroy ember views in destroy-render-node hook

The way `clearRender` works, it won’t call destroy on the top level
view if given `destroySelf` flag set to false. By moving the
`view.destroy` call into this hook, we can avoid destroying the top
level view when it is being cleared by `clearRender`.

The `view.lastResult` check seemed to just happen to work and needed to
be changed to get more tests passing.

mmun committed:
Merge pull request #10924 from mitchlloyd/view-remove-fixes

Get more tests with view#remove working

mixonic committed:
Tweak childNodes to fix test

mixonic committed:
Remove unnnecessary test

mixonic committed:
Revert incorrectly bumped htmlbars in a79c5868

mixonic committed:
Use unbound to assert dom rerender

mitchlloyd committed:
Allow attr names with `:` in attributeBindings

mixonic committed:
Throw when rerender called after rendering

Just before a block is rendered, views are put into a renderedViews
list. When `rerender` is called, this list is checked to ensure the view
being rerendered has not already been rerendered.

mitchlloyd committed:
Add assertion on 'class' for attributeBindings

mitchlloyd committed:
Update number of observers in test

It seems that this test was passing in spirit: the number of observers
does not increase when rerendering.

mitchlloyd committed:
Disallow spaces in classNameBindings property

Interestingly, spaces are allowed when setting classNameBindings on a
view inside of a template.

{{view classNameBindings="prop:true:false anotherProp:more-true:more-false"}}

However, they are not allowed when setting the property on a view class:

Ember.View.create({
  classNameBindings: 'prop:true:false anotherProp:more-true:more-false'
});

To handle both of these cases, we split the incoming classNameBindings
property by spaces when using the view keyword.

mixonic committed:
Merge pull request #10957 from mitchlloyd/view-attr-fixes

View attribute fixes

mixonic committed:
Fix layout tests

rwjblue committed:
Update HTMLBars to 0.13.8.

mixonic committed:
Update template_test to use compiled templates

mixonic committed:
Port deprecated tests, add test for passing attrs prop to component

mitchlloyd committed:
Add back assertion for tagless components

mixonic committed:
Fix arguments to handlebars compat helpers, and render return values

rwjblue committed:
Make proper test for makeViewHelper being looked up lazily from container.

rwjblue committed:
Merge pull request #10960 from mitchlloyd/tagless-assertion

Add back assertion for tagless components with classNameBindings

machty committed:
Get remainder of query_params_test.js passing

And a link-to test

rwjblue committed:
Merge pull request #10969 from machty/ir-qp

Get remainder of query_params_test.js passing

mmun committed:
Only proxy through dependencies when necessary, and make it painfully obvious

mmun committed:
Refactor SubexprStream

mmun committed:
Differentiate between mutable and immutable dependencies

this.addDependency now assumes that the dependee is immutable. In
particular this means that it no longer returns the Dependency instance.
If you need access to the dependency you must now call
this.addMutableDependency.

This change allows this.addDependency to be a no-op if the passed in
object is not a stream. This is useful for subexpression streams to
avoid creating unnecessary dependencies for primitive values, as in
this example:

```hbs
{{concat values sep=','}}
```

mmun committed:
Rename SimpleStream to ProxyStream

tilde-engineering committed:
Resolve master merge conflicts

tilde-engineering committed:
Merge branch 'idempotent-rerender' of http://github.com/emberjs/ember.js into idempotent-rerender

tilde-engineering committed:
Merge branch 'master' into idempotent-rerender

Commit:
publish glimmer builds to bower

wycats committed:
Initial incomplete impl of mut bindings

tomdale committed:
Support re-rendering inside lifecycle hooks

tomdale committed:
Bump htmlbars version

tomdale committed:
Unskip some link-to tests

tomdale committed:
Assert when calling view helpers with params

wycats committed:
Add support for intercepting get and set

These hooks are internal-only, not exposed publicly, and are only
intended to be used to emulate deprecated functionality for the moment.

tomdale committed:
Fix incomplete query params support in LinkView

mmun committed:
Fix preventDefault on link-to

mmun committed:
Recover willCreateElement hook to support instrumentation

mmun committed:
Fix controller usage in {{action}} and sendAction

mmun committed:
Bump emberjs-build to 0.1.1

krisselden committed:
Merge branch 'master' into idempotent-rerender

Conflicts:
	packages/ember-views/lib/attr_nodes/legacy_bind.js
	packages/ember-views/lib/mixins/class_names_support.js
	packages/ember-views/lib/streams/should_display.js
	packages/ember-views/lib/views/collection_view.js
	packages/ember-views/lib/views/core_view.js
	packages/ember-views/lib/views/select.js

rwjblue committed:
Merge pull request #10990 from krisselden/idempotent-rerender

Idempotent rerender

raytiley commented:
Just built latest code from this branch and `core_view` is throwing tons of `utils.typeOf is not a function`. Is this the best way to report issues when trying out this branch?

```js
 has: function (name) {
      return utils.typeOf(this[name]) === "function" || this._super(name);
    }
```

rwjblue committed:
Fix lint errors.

* `get` was unused in ember-routing-htmlbars/keywords/action
* Indentation was incorrect in ember-htmlbars/tests/integration/mutable_binding_test

rwjblue committed:
Fix incorrect `typeOf` import.

`typeOf` moved from `ember-metal/utils` to `ember-runtime/utils` as of
https://github.com/emberjs/ember.js/pull/10988.

rwjblue commented:
@raytiley - Thanks for reporting, was introduced in #10990 and fixed in a2d9a41.

raytiley commented:
I still don't think `actions` are being dispatched correctly. I have the following template for a `schedule` route.

```htmlbars
{{#schedule-drag-canvas class="clearfix"}}
	<div class="schedule-main">
		{{outlet}}
	</div>

	{{#right-side-drawer class="schedule-side-pane"}}
		{{outlet 'schedule-sidebar'}}
	{{/right-side-drawer}}

	{{schedule-drag-helper}}
{{/schedule-drag-canvas}}
```

Actions happening inside the default `{{outlet}}` are raising an error that the `schedule-drag-canvas` component does not handle the action. That action should not be going to the`schedule-drag-canvas` it should be going to `ScheduleController` and then bubbling up the route hierarchy where I handle it in `ScheduleRoute`.

chadhietala commented:
It appears file size gzipped is about 9kb larger then the 1.11.2 and 8kb from 1.12.0-beta.  Not sure if this is to be expected.

rwjblue commented:
@chadhietala - Canary builds are always larger than stable/beta builds due to the feature flag process (all flagged code is still included in canary builds but is completely stripped from beta/stable).  Just how much the difference is after that I am unsure...

chadhietala commented:
@rwjblue Ah yes, forgot about defeatureify. I should know this by now.

ef4 commented:
I also wouldn't get too excited about the size falling until 2.0. Glimmer is fully backward compatible still. It's at 2.0 that we get to drop cruft.

mitchlloyd committed:
Notify view controller when its parentView changes

tomdale committed:
Set controller inside {{#each arrayController}}

tomdale committed:
Refactor controller binding

tilde-engineering committed:
Preserve controller instance across rerenders

tilde-engineering committed:
Allow {{with}} to bind controller

rwjblue committed:
Add integration test for makeViewHelper looked up from the container.

wycats committed:
Mostly complete impl of mutable bindings

There are two failing tests related to timing and I want to investigate
an alternate approach for mutable cells.

wycats committed:
Don't leak the stream into user code

In the previous commit, mutable bindings had a value() method that was a
live reflection of the underlying stream. Those semantics were
inconsistent with immutable bindings, which only update during the
rendering process and trigger lifecycle hooks.

This commit snapshots the value and provides a new mutable cell each
time through.

mitchlloyd committed:
Consistently use `parentView` over `_parentView`

rwjblue committed:
Merge pull request #10999 from mitchlloyd/fix-create-child-view-tests

Consistently use `parentView` over `_parentView`

rwjblue committed:
Merge pull request #10991 from mitchlloyd/view-computed-controller

Fix a test for view controller property

blesh commented:
FYI: I was speaking with @stefanpenner last night about some perf issues we had, and I put together this example: https://github.com/blesh/is-ember-really-fast-yet

Just pull it down and run it

I didn't include a baseline, but we have one here, it's complicated because it's all D3 but it runs in about 1 sec (as opposed to ~16secs)

I hope this is helpful.

bantic committed:
Fallback to "-text-field" component name for unrecognized input type

Previously, using an input type not in the map (such as "email") would
cause an exception. This renames the `classification` var to more
expressive `componentNameMap`, and turns the map into a whitelist of
non-default component names only. Any value of "type" not in that map (or
no value for "type") will use the default "-text-field" component name.

wycats committed:
Merge pull request #11004 from bantic/idempotent-rerender-fix-input-types

Fallback to "-text-field" component name for unrecognized input type

rwjblue committed:
Use attrs by default and fallback to root properties in TextSupport.

rwjblue committed:
Use modern template in ember-testing integration tests.

rwjblue committed:
Always set `elementId`.

The container view template uses `elementId` as the key for `{{each}}`.
We need to ensure that even tagless views have an `elementId`.

rwjblue committed:
Update remaining skipped `View.prototype.createElement`.

* Avoid using `RenderBuffer`
* assert the output HTML

rwjblue committed:
Prevent clobbering attrs with an attr named attrs.

When invoking a component like:

```javascript
{{foo-bar attrs="blah"}}
```

We cannot set `this.attrs` to be equal to `"blah"`.  This is potentially
breaking for folks that have used `attrs` as a single attribute value in
the root of a component.

wycats committed:
Merge pull request #10958 from mixonic/idempotent-rerender-assert-rerender

Throw when rerender called after rendering

rwjblue committed:
Use `indexOf` from EnumerableUtils.

It properly defers to the object or polyfil as needed.

tilde-engineering committed:
Add failing controller scope/action test

tilde-engineering committed:
Rename componentNode -> manager

tomdale committed:
Extract simplified ComponentNodeManager

tomdale committed:
Extract lookupComponent

mmun committed:
Make streams aware of proxies

If a stream value results in a proxy, the stream will begin to observe
the proxy's content and notify itself if the content changes.

krisselden committed:
Merge branch 'master' into idempotent-rerender

rwjblue committed:
Merge pull request #11008 from krisselden/idempotent-rerender

Merge master into idempotent rerender

mixonic committed:
Style cleanup to fix tests

wycats committed:
Fix issue with inputs getting re-rendered

rwjblue committed:
Bump HTMLBars version to 0.13.10.

rwjblue committed:
Fix failed teardown in `View.prototype.appendTo` tests.

wycats committed:
Fixes input getting re-rendered incorrectly

wycats committed:
Adds a way to reliably read from DOM

The readDOMAttr method on views makes it possible to read an attribute
from DOM and update the internal lastValue at the same time.

This effectively fixes the fact that some attribute sets are not
idempotent, which causes problems for the write-only pure-function
rendering model used by Glimmer and React.

rwjblue committed:
Fix dynamic `makeViewHelper`.

rwjblue committed:
Fix dynamic `makeViewHelper`.

wycats commented:
@blesh 1s instead of 16s sounds better? Or did I misunderstand?

blesh commented:
@wycats ... I was saying it currently takes 16 seconds to render 60 graphs with an {{#each}}... @stefanpenner helped me find that 10 seconds of that was the graph component. But the remaining 6 is still pretty fair away from the baseline of ~1 second if things are done in pure d3.

blesh commented:
@wycats, after updates and testing, it was a negligible amount that was actually ember-nf-graph. There were perf issues, for sure, but fixing them didn't make a huge difference.

rwjblue committed:
Use readDOMAttr + attributeBinding for <textarea>.

The new `readDOMAttr` API allows us to remove this manual `value`
property binding in `Ember.TextArea`.

rwjblue committed:
Update HTMLBars to v0.13.11.

* Remove 'value' attribute hacks.
* Add `AttrMorph.prototype.getContent`.

rwjblue committed:
Remove unused imports in Ember.TextArea.

rwjblue committed:
Merge pull request #10977 from fivetanley/bower-idemptotent-rerender

publish idemptotent rerender to bower

mixonic committed:
Bring back the render hook with a pushable buffer

mmun committed:
Fix stream teardown... I think

rwjblue committed:
Set outlet state prior to rendering {{outlet}} in tests.

mixonic committed:
Merge pull request #11012 from mixonic/idempotent-rerender-buffer

Bring back the render hook with a pushable buffer

tilde-engineering committed:
Propagate controller keyword for outlets

mixonic committed:
Drop old render_buffer file in favor of compat/ file

tilde-engineering committed:
Fix {{#each}} with itemController in outlet

mixonic committed:
Skipped test for values propagating upstream via templates

tilde-engineering committed:
Fix controller local w/ `{{#each itemController}}`

rwjblue committed:
Make JSCS happy.

mixonic committed:
Add a (passing?) test for settable upstreaming template bindings

rwjblue committed:
Fix propagation test.

mmun committed:
Intercept property changes instead of sets

(cherry picked from commit b4d53bfd2d2769c4133f5f55e5551c6cc2dc87f8)

rwjblue committed:
Remove deprecation on accessing `attrs` in component root.

We have not fully fleshed out the transition plan for this. Remove the
deprecation until we have...

bcardarella commented:
ZOMG

raytiley commented:
Getting an error pretty regularly about `morph.clear()` is not a function. Not sure if its related the todo in the code. I'm not exactly sure how you get a logical child that's not in the dom.

![windows 7 x64 2015-05-04 10-00-26](https://cloud.githubusercontent.com/assets/41680/7454779/fd6362d8-f244-11e4-80f1-72cd977ae73f.png)

rwjblue committed:
Add skipped test showing issue with mutable binding system.

wycats committed:
Don't warn when for-compat stream is not a path

The `(mut foo)` construction is designed to throw an error if you try to
do `(mut "foo")`, for obvious reasons.

When I wrote that error, I missed an alternative scenario:

```hbs
{{!-- component-one: --}}

{{!-- bar is not a path here --}}
<component-two foo="bat" />

{{!-- component-two: --}}

{{!-- attrs.foo is a path here --}}
<component-three baz={{mut attrs.foo}} />
```

In this case, `attrs.foo` is a path, but it directly refers to a string,
which is itself not mutable.

With angle-bracket components, when `{{mut}}` is explicit, the error
still makes sense: `<component-three>` should not try to mutate
`attrs.foo` and we should warn as soon as we can.

Unfortunately, that logic does not apply to the compatibility work we
are doing:

```hbs
{{!-- component-one: --}}

{{component-two foo="bat"}}

{{!-- component-two: --}}

{{component-three baz=attrs.foo}}
```

In this case, we can't know ahead of time that `attrs.foo` refers to a
non-mutator, so we have to allow for the possibility that this is a
two-way binding. For what it's worth, scenarios like this, which are
clearly bugs (`component-three` settting `baz`), help to explain the
motivation for the change in the first place.

This commit adds a private `@mut` that can support the looser semantics
of Ember 1.x curly components, while retaining the stricter errors for
explicit calls to `mut` used with angle bracket components.

rwjblue committed:
Unskip mutable binding test.

rwjblue committed:
Publish Glimmer builds to S3.

Make testing apps easier...

rwjblue committed:
Allow publishing of idempotent-rerender PR to S3.

rwjblue committed:
Bring Ember._Metamorph and Ember._MetamorphView back and deprecate.

ember-test-helpers still attempts to register `Ember._Metamorph` as
`view:default` which causes an error. We will need to update
ember-test-helpers to check for its existence first and remove the
globals at a later date.

rwjblue committed:
Revert "Allow publishing of idempotent-rerender PR to S3."

This reverts commit 12b249632c4047c312441b69a7d51584595141dc.

This was accidentally pushing the glimmer builds to `canary` URL's.

Removed for now.  Lets just merge this PR....

rwjblue commented:
We have 3 tests still skipped at the moment (the fixes are in various stages of in-flight). I recommend that we merge, and iterate on the remaining issues in canary.  To make this a nice stable beta (in roughly a week) we need more users to be able to test these changes in their applications and report any issues they come across.

@emberjs/owners - Any objections?

wycats commented:
@rwjblue works for me. I plan to have a blog post explaining the status and remaining open issues in the morning (Greek time).

chadhietala commented:
:shipit: Will hook it up to an app this week.

blesh commented:
So I ran `1.12.0-beta.1+canary.27ed45f8` against https://github.com/blesh/is-ember-really-fast-yet

Unfortunately, I didn't see much improvement for my use case of graphs in an each. Most of the time spent is in `scheduleRender`, which is a little hard to see in the production build profile. 

Here's what the production build profile looks like:

![screen shot 2015-05-04 at 4 25 14 pm](https://cloud.githubusercontent.com/assets/1540597/7464729/419824d6-f27a-11e4-8132-66214676341f.png)


It takes about 8-9 seconds to render, as opposed to a baseline of ~1sec. (You can see the baseline if you clone the repo and run it, navigating to http://localhost:4200/basline, thanks to @jeff3dx)

Graphs in an each weren't the only place I've seen this particular issue. Basically any large number of views or components in an each will cause this slowness.

I have the .cpuprofile file if you're all interested.

cc/ @stefanpenner

mixonic committed:
outlet keywords must pass the template to render as template

Previously, they were passing it as layout which smashed a layout that
may have existed on the view being rendered.

rwjblue committed:
Merge pull request #11022 from mixonic/idempotent-rerender-layout

outlet keywords must pass the template to render as template

rwjblue commented:
@blesh:

This refactor addresses re-render performance by avoiding creating, destroying, and re-recreating DOM elements and views/components. It does improve the cost of view/component instantiation, but not significantly (which I would suspect is a fairly significant factor in the case you reference).  There are a number of contributing factors to the initial view render (as a few others have mentioned), and this PR is definitely a step in the right direction.

This refactor along with some Ember 2.0 spring cleaning and straight up performance work, will address the initial render time speed issues your examples are referencing.

mmun committed:
Fix a memory leak when KeyStream is observering a proxy

mmun committed:
Remove dead code

mmun committed:
Privatize bind-attr class compat helper

mmun committed:
Remove more dead code, relocate docs

sandstrom commented:
@blesh I feel this pain too, but I think it's orthogonal to glimmer. [This dedicated issue](https://github.com/emberjs/ember.js/issues/10261) is better to track the case you mention.

I've created a test-repo without the graph component (easier to investigate performance without third-party code), running on Glimmer (even though I don't think that has much of an effect, for the reasons @rwjblue mentions).

rwjblue committed:
Add test controller propagation from route driven views.

(cherry picked from commit 035545ed2c47b84f1c104d3e22bded6f72db1a37)

rwjblue committed:
Ensure that `controller` is provided for route backed views.

Fixes #11013.

bcardarella commented:
![](http://38.media.tumblr.com/tumblr_lunocz3mMX1r24lq9o1_250.gif)

Emerson commented:
Congrats on the merge - been following along since Ember Conf - super exciting stuff!

rondale-sc commented:
![](http://media.tumblr.com/e4dac0bdf94be16226d71eb91ea04fe6/tumblr_inline_mj9ynvgRJu1qz4rgp.gif)

listrophy commented:
![clapping](https://cloud.githubusercontent.com/assets/26813/7476005/967ccb32-f310-11e4-961e-26ac37a82cab.gif)

rtablada commented:
![There was much rejoicing](https://bostonstitletowntalk.files.wordpress.com/2014/07/rejoicing.gif)

bcardarella commented:
![](http://i.imgur.com/VLm02E6.gif)

danmcclain commented:
![tumblr_m3vi749jwv1r3zat8](https://cloud.githubusercontent.com/assets/144138/7476135/a42222f6-f319-11e4-9947-cac7d93c300a.gif)

jdjkelly commented:
![005385766_2cool](https://cloud.githubusercontent.com/assets/704789/7476181/f52243a2-f319-11e4-8b1a-0abb5274bdb5.gif)

JPBetley commented:
![](http://media3.giphy.com/media/M7gtacN7aPNsc/giphy.gif)

chadhietala commented:
![](http://media.giphy.com/media/fcuyIsxkuKRFe/giphy.gif)

thejameskyle commented:
*googles "glimmer gif"* [](http://media.tumblr.com/b2224c4ba1a26dff4663e61249fcdb4d/tumblr_inline_mhy9grcKMM1ropvqs.gif)

rtablada commented:
@thejameskyle results may vary

jdalton commented:
![woooooooo!](https://cloud.githubusercontent.com/assets/4303/7476439/5d110740-f302-11e4-882e-adb69a5dc1f9.gif)

MattMSumner commented:
![i_love_it](https://cloud.githubusercontent.com/assets/759721/7476674/c532545c-f314-11e4-93f5-2b063109b4ca.gif)

jeanleonino commented:
Amen!

bartocc commented:
Congrats !

Adriaaaaan commented:
Anyone got it working yet?  I just get 
> Uncaught TypeError: Cannot read property 'forEach' of undefined

childviews doesn't exist

```javascript
    appendChildren: function(view, children, retained) {
        var self = this;
        var childViews = view.get('_childViews');

        childViews.forEach(function(childView) {
          if (!(childView instanceof Ember.Object)) { return; }

          if (self.shouldShowView(childView)) {
            var grandChildren = [];
            children.push({ value: self.inspectView(childView, retained), children: grandChildren });
            self.appendChildren(childView, grandChildren, retained);
          } else {
            self.appendChildren(childView, children, retained);
          }
        });
      },
```

mixonic commented:
@Adriaaaaan `_childViews` is a private API that has been removed. Please use `view.get('childViews')`.

WillsonSmith commented:
![eznaclq](https://cloud.githubusercontent.com/assets/1087756/7480566/69e7b90c-f337-11e4-821d-ea9ea5857102.gif)

bdvholmes commented:
Woohoo!

oskarrough commented:
Just updated from 1.11.1 and it took five minutes to get everything working through all your console.logs. So good!

mrinterweb commented:
![](http://i.imgur.com/vwwTGp7.gif)

Fed03 commented:
![image](http://i.giphy.com/xPGkOAdiIO3Is.gif)

tubbo commented:
http://i.imgur.com/epVWJFE.gif

btecu commented:
Great work!

Adriaaaaan commented:
Great work! although our app is currently very broken still I'm already seeing a huge improvement in performance in my list views.  Now i just have a million deprecation warnings to fix

wycats commented:
@Adriaaaaan That's great news! Please do report any notable breakage. We want things to be pretty clean by the time we ship 1.13 final :)

You might also want to run in prod mode to check out perf. Those deprecation warnings aren't free to generate ;)

chrism commented:
Great Work!

Anyone using Liquid Fire should check that https://github.com/ef4/liquid-fire/issues/276 has been resolved before updating as it is not yet glimmer compatible.

jmurphyau commented:
@wycats any help you can provide on unbound helpers? A helper that produces a stream like value that can be invalidated causing the value in the template to update? Trying to get [ember-get-helper](https://www.npmjs.com/package/ember-get-helper) to work with glimmer.. Simply returning a new child stream (e..g `objectStream.get(key)`) doesn't cause a new value to render in the template when that value changes.. Do I need to look into some of the hooks or somehow subscribe <some object> to the new stream?

jmurphyau commented:
I can't seem to find that really detailed description of all the HTMLBars lifecycle hooks etc.. Do you have a link to that that you could provide?

wycats commented:
@jmurphyau right now there is no public way for a helper to express dynamic dependencies, but that's a good idea. 

Glimmer intentionally doesn't expose the internal stream machinery; it's changing quite a bit still, but it makes sense for a helper to be able to tell Ember when to invalidate it. 

It also kind of makes sense for us to include this get helper in Ember itself honestly. It's a primitive, like `{{component}}`.

wycats commented:
@jmurphyau as I said in the blog post, more info on the life cycle hooks is forthcoming :)

Adriaaaaan commented:
@wycats Hmmm have looked into it more, I'm now seeing a rather significant performance regression.  related to interacting with sub resources templates from a list.  It seems that interacting with it causes the the whole tree to redraw making it unusable if there are a lot of rows per page.  1.11 is faster, its most notable when selecting the rows (the detail page is very laggy at appearing, the more rows in the list the longer it takes).  Although it is clearly much faster at rendering (the page appears much faster) interacting with it is much worse as it seems to be doing more work everytime something changes.  I'll see If I can make a simple jsbin

EnglishCriminal commented:
![](http://media.giphy.com/media/MUeQeEQaDCjE4/giphy.gif)

jiyinyiyong commented:
Hi, you mentioned there's slide in EmberConf 2015. But is there a video?
I can't find it in the list "EmberConf 2015 by Confreaks"
https://www.youtube.com/playlist?list=PLE7tQUdRKcyacwiUPs0CjPYt6tJub4xXU

igorT commented:
@jiyinyiyong it's the keynote, towards the end

jiyinyiyong commented:
@igorT I don't get it. Could you point it out in the video?

blessenm commented:
Stupid question. So there is no virtual DOM in ember right? Just value diffing to find which part of the DOM needs to be manipulated.

axemclion commented:
Glimmer is pretty fast - almost 25% faster !! Ran some tests and results here - http://blog.nparashuram.com/2015/05/performance-boost-in-emberjs-from.html

jdurand commented:
@blessenm instead of diffing DOM it diffs a tree of possibly mutable states that are bound to the real DOM. It doesn't need to maintain a virtual DOM because it has deeper knowledge of your apps state.

blessenm commented:
@jdurand thanks for the clarification.

ballPointPenguin commented:
![The Glimmer Twins](http://i146.photobucket.com/albums/r249/mikegryn/mkglimmer.jpg)

jmurphyau commented:
@wycats someone has asked a relatively simple question on [stackoverflow](http://stackoverflow.com/questions/30776252/streams-in-ember-js) and in [slack](https://embercommunity.slack.com/archives/random/p1434100649001108) - 'what is a stream'.. A answered as best I could but wanted to bring it up here - since this is where he got it from - what is a stream?

matthewp commented:
Is there somewhere that explains how Glimmer works in more detail, with links to code maybe? Given how many commits are included in this PR it's hard to figure out how the details are implemented in practice.  For example you say:

> On initial render, we track the last value (always a primitive) that we inserted into the DOM.

But I can't find in the code where this happens. And I don't know what you mean by a primitive value.  I assume you mean that you track attribute string values and text node's string nodeValues but am having a difficult time finding where this happens in the code.

Is there maybe one or two commits that contain most of the work that I can look at?

ef4 commented:
@matthewp Much of the infrastructure is actually in https://github.com/tildeio/htmlbars

There's not really a single commit to point to -- the work was extensive and many people put in hundreds of commits.
